src/lustre/lustreTypeChecker.ml

```diff
@@ -242,8 +242,229 @@ let extract_consts: LA.const_clocked_typed_decl -> tc_context
+  eval_bool_expr ctx e2 >>= fun v2 ->
+  match bop with
+  | And -> R.ok (v1 && v2) 
+  | Or -> R.ok (v1 || v2)
+  | Xor -> R.ok ((v1 && not v2) || (v2 && not v1))
+  | Impl -> R.ok (not v1 || v2)
+  | _ -> type_error pos ("Cannot evaluate non-bool binary expression"
+                         ^ string_of_expr (LA.BinaryOp (pos, bop, e1, e2))
+                         ^" to a bool value")
+(** try and evalutate binary op expression to bool, return error otherwise *)
   
+and eval_bool_ternary_op: tc_context -> Lib.position -> LA.ternary_operator
+                     -> LA.expr -> LA.expr -> LA.expr -> bool tc_result
+  = fun ctx pos top b1 e1 e2 ->
+  eval_bool_expr ctx b1 >>= fun c ->
+  eval_bool_expr ctx e1 >>= fun v1 ->
+  eval_bool_expr ctx e2 >>= fun v2 ->
+  match top with
+  | LA.Ite -> if c then R.ok v1 else R.ok v2
+  | LA.With -> type_error pos "With operator is not supported"
+(** try and evalutate ternary op expression to bool, return error otherwise *)
+
+and eval_int_ternary_op: tc_context -> Lib.position -> LA.ternary_operator
+                     -> LA.expr -> LA.expr -> LA.expr -> int tc_result
+  = fun ctx pos top b1 e1 e2 ->
+  match top with
+  | LA.Ite ->
+     eval_bool_expr ctx b1 >>= fun c ->
+     if c
+     then eval_int_expr ctx e1
+     else eval_int_expr ctx e2
+  | LA.With -> type_error pos "With operator is not supported"
+(** try and evalutate ternary op expression to int, return error otherwise *)
+
+             
+and eval_comp_op: tc_context -> Lib.position -> LA.comparison_operator -> LA.expr -> LA.expr -> bool tc_result = 
+  fun ctx pos cop e1 e2 ->
+  eval_int_expr ctx e1 >>= fun v1 ->
+  eval_int_expr ctx e2 >>= fun v2 ->
+  match cop with
+  | Eq -> R.ok (v1 = v2)
+  | Neq -> R.ok (v1 <> v2)
+  | Lte -> R.ok (v1 <= v2)
+  | Lt -> R.ok (v1 < v2)
+  | Gte -> R.ok (v1 > v2)
+  | Gt -> R.ok (v1 >= v2)
+(** try and evalutate comparison op expression to bool, return error otherwise *)
+
+and simplify_array_index: tc_context -> Lib.position -> LA.expr -> LA.expr -> LA.expr
+  = fun ctx pos e1 idx -> 
+  match (simplify_expr ctx e1) with
+  | LA.GroupExpr (pos, ArrayExpr, es) ->
+     (match (eval_int_expr ctx idx) with
+      | Ok i -> if List.length es > i
+                then List.nth es i
+                else
+                  (Log.log L_fatal "Trying to access element out of bounds %a"
+                     Lib.pp_print_position pos
+                  ; raise OutOfBounds)
+      | Error _ -> LA.ArrayIndex (pos, e1, idx))
+  | _ -> ArrayIndex (pos, e1, idx)
+(** picks out the idx'th component of an array if it can *)
+
+and simplify_tuple_proj: tc_context -> Lib.position -> LA.expr -> int -> LA.expr
+  = fun ctx pos e1 idx ->
+  match (simplify_expr ctx e1) with
+  | LA.GroupExpr (pos, _, es) ->
+     if List.length es > idx
+     then List.nth es idx
+     else (Log.log L_fatal "Trying to access element out of bounds %a"
+             Lib.pp_print_position pos
+          ; raise OutOfBounds)
```

Comment:
Debug messages are the only kind of message this module should log, not error messages. You should catch this exception within this module (type_check_program), and return an Error.